### PR TITLE
[enh] Add timeout limit per request

### PR DESCRIPTION
**Pull request by dalf**
The new url parameter "timeout_limit" set timeout limit defined in second.
Example ```"timeout_limit=1.5"``` means the timeout limit is 1.5 seconds

In addition, the query can start with <[number] to set the timeout limit.

For number between 0 and 9, the number unit is the second :
Example: ```"<3 searx"``` means the timeout limit is 3 seconds

For number above 9, the number unit is the centisecond:
Example: ```"<150 searx"``` means the timeout is 1.5 seconds

Related to #1077
Updated version of PR #1413 from @isj-privacore

### CODE REVIEW

**Comment by Pofilo** [comment]
@dalf is it still in WIP ? Can I read and test it ?

**Comment by dalf** [comment]
@Pofilo : you can read / test it ! Thank you !

File: searx/query.py [hunk 1]
```diff
@@ -69,6 +70,19 @@ def parse_query(self):
                 self.query_parts.append(query_part)
                 continue
 
+            # this force the timeout
+            if query_part[0] == '<':
+                try:
+                    raw_timeout_limit = int(query_part[1:])
+                    if raw_timeout_limit < 10:
```
Comment (Pofilo) [inline searx/query.py]: Maybe adding a comment about seconds and centiseconds could help for the future.

**Comment by Pofilo** [comment]
I don't have enough time these days to deploy a platform for testing, however, I don't have anything to say about the core of the feature. Feel free to add the comments if you want (only if you have time for this). Do I need to test it or your test is enough according to you ?

**Comment by dalf** [comment]
@Pofilo no problem. 

Update:
* bug fix: get_search_query_from_webapp ignored "<[num]" if the URL query parameter "timeout_limit" was not specified
* the real timeout limit is minimum between the one ask by the user and the normal timeout. Purpose: avoid possible deny of service ask for a lot of queries with a long timeout limit.
* ```<[anything else than a number]``` will be processed as part of the search query (similar behavior to other special characters processing ```!```, ```?```, ```:```)

**Comment by dalf** [comment]
I changed the bang ```<[num]``` to something perhaps more intuitive:
* below 100, the unit is the second: ```<30``` = 30 seconds timeout.  
* 100 or above, the unit is the millisecond: ```<850``` = 850 milliseconds timeout.

More over, there is a new optional value in settings.yml: ```outgoing.max_request_timeout```.

engine timeout | timeout_limit query param(*) | settings.yml... max_request_timeout | Actual timeout | Comment
--: | --: | --: | --: | :--
3 | None | None | 3 |  
3 | 2 | None | 2 |  
3 | 5 | None | 3 | timeout_limit is ignored
3 | None | 30 | 3 |  
3 | 2 | 30 | 2 |  
3 | 5 | 30 | 5 | timeout_limit is NOT ignored

(*) either the timeout_limit query parameter or the bang ```<[num]``` (if both are specified, priority for the latter one)

Why this complicated thing:
* for people who don't set max_request_timeout, no surprise, the timeout won't go above the previous values (engine timeout).
* for people who set a max_request_timeout, timeout_limit can use the full range from 0 to max_request_timeout.

[EDIT] explain priority between the timeout_limit query parameter and ```<[num]```

**Pull request merged**